src/unix/udp.c

```diff
@@ -616,23 +639,19 @@ static int uv__setsockopt_maybe_char(uv_udp_t* handle,
   if (val < 0 || val > 255)
     return -EINVAL;
 
-  if (handle->flags & UV_HANDLE_IPV6)
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IPV6,
-                   option6,
-                   &arg,
-                   sizeof(arg));
-  else
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IP,
-                   option4,
-                   &arg,
-                   sizeof(arg));
-
-  if (r)
-    return -errno;
+/*
+ * On Solaris and derivatives such as SmartOS, the length of socket options
+ * is always sizeof(int), except for the IP_MULTICAST_LOOP and IP_MULTICAST_TTL
+ * options.
+ */
+#if defined(__sun)
```

Comment:
I think I'd rather see this exception on each case, instead of the general one:

~~~~
int uv_udp_set_multicast_ttl(uv_udp_t* handle, int ttl) {
#if defined(__sun)
  if ((handle->flags & UV_HANDLE_IPV6)
    uv__setsockopt(handle, IP_MULTICAST_TTL, IPV^_MULTICAST_HOPS, ttl, sizeof(ttl));
#endif 
  return uv__setsockopt_maybe_char(handle,
                                   IP_MULTICAST_TTL,
                                   IPV6_MULTICAST_HOPS,
                                   ttl);
}
~~~~

and the same for uv_udp_set_multicast_loop.